3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala

```diff
@@ -0,0 +1,104 @@
+ * @param timeout max duration of each connection attempt, and of the wait for each response from the remote
+ */
+class RemoteAdapter(val remoteUrl: String, val timeout: FiniteDuration) extends Adapter {
+
+  val bodyMissingErrorText = "missing payload body"
+
+  private lazy val log = LoggerFactory.getLogger(getClass)
+
+  val remoteUri = uri"$remoteUrl"
+  implicit val backend = HttpURLConnectionBackend(
+    options = SttpBackendOptions.connectionTimeout(timeout)
+  )
+
+  /**
+   * POST the given payload to the remote webservice,
+   * wait for it to respond with an Either[List[String], List[RawEvent] ],
+   * and return that as a ValidatedRawEvents
+   *
+   * @param payload The CollectorPaylod containing one or more
+   *        raw events as collected by a Snowplow collector
+   * @param resolver (implicit) The Iglu resolver used for
+   *        schema lookup and validation
+   * @return a Validation boxing either a NEL of RawEvents on
+   *         Success, or a NEL of Failure Strings
+   */
+  def toRawEvents(payload: CollectorPayload)(implicit resolver: Resolver): ValidatedRawEvents =
+    if (payload.body.isEmpty || payload.body.get == "") {
+      bodyMissingErrorText.failNel
+    } else {
+
+      val json = ("contentType" -> payload.contentType) ~
+        ("queryString" -> toMap(payload.querystring)) ~
+        ("body"        -> payload.body)
+
+      httpPost(remoteUrl, compact(render(json))) match {
+        case Left(errmsg) =>
+          errmsg.failNel
+        case Right(bodyAsString) =>
+          try {
+            val error = (parse(bodyAsString) \ "error" ).extract[String]
+            if (error != null) {
+              error.failNel
+            } else {
+              val events = (parse(bodyAsString) \ "events" ).extract[List[Map[String, String]]]
+              rawEventsListProcessor(events.map { event =>
+                RawEvent(
+                  api         = payload.api,
+                  parameters  = event,
+                  contentType = payload.contentType,
+                  source      = payload.source,
+                  context     = payload.context
+                ).success
+              })
+            }
+          } catch {
+            case  e: Exception => s"Unexpected response from remote $e".failNel
+          }
+
+
+        case ng => s"Unexpected response from remote $remoteUrl $ng".failNel
+      }
+    }
+
+  private def httpPost(requestUrl: String, body: String): Either[String, String] =
+    try {
+      val request  = sttp.readTimeout(timeout).body(body).post(uri"$requestUrl")
+      val response = request.send()
+
+      if (response.isClientError)
+        Left(s"remote webservice rejected this request with a ${response.code} error! ('${response.body}')")
+      else if (response.isServerError)
+        Left(s"remote webservice choked on this request, got a ${response.code} error! ('${response.body}')")
+      else
+        Right(response.unsafeBody)
+
+    } catch {
```

Comment:
Using `Try` and pattern matching would be nicer than having a big `try {} catch {}` and `if () else if () else ()`.
```scala
Try(sttp.readTimeout(timeout).body(body).post(uri"$requestUrl"))
  .flatMap(_.send()) match {
    case Success(response) if response.isClientError => ...
    case Success(response) if response.isServerError => ...
    case Success(response) => ...
    case Failure(err) => ...
  }
```